hack/generator/pkg/conversions/hub_conversion_function.go

```diff
@@ -0,0 +1,276 @@
+	}
+
+	conversionPackage := generationContext.MustGetImportedPackageName(astmodel.ControllerRuntimeConversion)
+
+	funcDetails.AddParameter("hub", astbuilder.QualifiedTypeName(conversionPackage, "Hub"))
+	funcDetails.AddReturns("error")
+	funcDetails.AddComments(fn.declarationDocComment(receiver))
+
+	if fn.intermediateType == nil {
+		funcDetails.Body = fn.DirectConversion(receiverName, generationContext)
+	} else if fn.direction == ConvertFrom {
+		funcDetails.Body = fn.IndirectConversionFromHub(receiverName, generationContext)
+	} else if fn.direction == ConvertTo {
+		funcDetails.Body = fn.IndirectConversionToHub(receiverName, generationContext)
+	} else {
+		panic(fmt.Sprintf("unexpected conversion direction %q", fn.direction))
+	}
+
+	return funcDetails.DefineFunc()
+}
+
+func (fn *HubConversionFunction) DirectConversion(
+	receiverName string, generationContext *astmodel.CodeGenerationContext) []dst.Stmt {
+	fmtPackage := generationContext.MustGetImportedPackageName(astmodel.FmtReference)
+
+	localId := fn.localVariableId()
+	localIdent := dst.NewIdent(localId)
+	hubIdent := dst.NewIdent("hub")
+
+	assignLocal := astbuilder.TypeAssert(
+		localIdent,
+		hubIdent,
+		astbuilder.Dereference(fn.hub.AsType(generationContext)))
+
+	checkAssert := astbuilder.ReturnIfNotOk(
+		astbuilder.FormatError(
+			fmtPackage,
+			fmt.Sprintf("expected %s but received %%T instead", fn.hub),
+			hubIdent))
+
+	copyAndReturn := astbuilder.Returns(
+		astbuilder.CallExpr(dst.NewIdent(receiverName), fn.propertyFunctionName, localIdent))
+
+	return astbuilder.Statements(assignLocal, checkAssert, copyAndReturn)
+}
+
+func (fn *HubConversionFunction) IndirectConversionFromHub(
+	receiverName string, generationContext *astmodel.CodeGenerationContext) []dst.Stmt {
+	errorsPackage := generationContext.MustGetImportedPackageName(astmodel.GitHubErrorsReference)
+	localId := fn.localVariableId()
+	errIdent := dst.NewIdent("err")
+
+	declareLocal := astbuilder.LocalVariableDeclaration(
+		localId, fn.intermediateType.AsType(generationContext), "// intermediate variable for conversion")
+
+	populateLocalFromHub := astbuilder.SimpleAssignment(
+		errIdent,
+		token.DEFINE,
+		astbuilder.CallExpr(dst.NewIdent(localId), fn.name, dst.NewIdent("hub")))
+
+	checkForErrorsPopulatingLocal := astbuilder.CheckErrorAndWrap(
+		errorsPackage,
+		fmt.Sprintf("converting from hub to %s", localId))
+
+	populateReceiverFromLocal := astbuilder.SimpleAssignment(
+		errIdent,
+		token.ASSIGN,
+		astbuilder.CallExpr(dst.NewIdent(receiverName), fn.propertyFunctionName, astbuilder.AddrOf(dst.NewIdent(localId))))
+
+	checkForErrorsPopulatingReceiver := astbuilder.CheckErrorAndWrap(
+		errorsPackage,
+		fmt.Sprintf("converting from %s to %s", localId, receiverName))
+
+	returnNil := astbuilder.Returns(dst.NewIdent("nil"))
+
+	return astbuilder.Statements(
+		declareLocal, populateLocalFromHub, checkForErrorsPopulatingLocal, populateReceiverFromLocal, checkForErrorsPopulatingReceiver, returnNil)
```

Comment:
minor: This would probably be clearer 1 line per arg

Reply:
Fixed here, and elsewhere calling `astbuilder.Statements()`